### PR TITLE
Fix #7281: Spinner: Horizontal mode

### LINKED ISSUE #7281: Spinner: Add horizontal buttons
So, something like

```
[-|_______|+]
```

or

```
[_______|-|+]
```

### PR DESCRIPTION
**Pull request by jepsar**
Fix #7281

Showcase https://github.com/primefaces/primefaces-showcase/pull/359
```xhtml
<p:outputLabel for="@next" value="Horizontal"/>
<p:spinner id="horizontal" value="#{spinnerView.number8}" buttons="horizontal"
           upIcon="pi pi-chevron-right" downIcon="pi pi-chevron-left"/>

<p:outputLabel for="@next" value="Horizontal after"/>
<p:spinner id="horizontalAfter" value="#{spinnerView.number9}" buttons="horizontal-after"
           upButtonStyleClass="ui-button-success" downButtonStyleClass="ui-button-danger"/>

<p:outputLabel for="@next" value="Vertical"/>
<p:spinner id="vertical" value="#{spinnerView.number10}" buttons="vertical"
           upButtonStyleClass="ui-button-secondary" downButtonStyleClass="ui-button-secondary"/>
```
Screenshot 2021-05-13 at 13 53 26

**Important!** This requires a change in the themes to fix the border radius and right padding. A new issue should be created when this is merged.

In order to get the custom icons working,
```css
body .ui-spinner .ui-spinner-button.ui-button.ui-spinner-xxx .ui-icon:before
```
should be changed to:
```css
body .ui-spinner.ui-spinner-stacked .ui-spinner-button.ui-button.ui-spinner-xxx .ui-icon:before
```
Or, we should add the default icons to the component and the rule above should be removed.

Styles that were added to this PR which should be moved to a theme:
```css
body .ui-spinner.ui-spinner-horizontal .ui-spinner-button.ui-button.ui-spinner-up,
body .ui-spinner.ui-spinner-horizontal-after .ui-spinner-button.ui-button.ui-spinner-up {
    border-radius: 0 4px 4px 0;
}

body .ui-spinner.ui-spinner-horizontal .ui-spinner-button.ui-button.ui-spinner-down {
    border-radius: 4px 0 0 4px;
}

body .ui-fluid .ui-spinner-horizontal-after .ui-spinner-input, body .ui-spinner-horizontal-after .ui-spinner-input {
    padding-right: 5.214rem;
}

.ui-spinner-horizontal-after .ui-spinner-down {
    right: 2.357rem;
}

body .ui-fluid .ui-spinner.ui-spinner-vertical .ui-spinner-input, body .ui-spinner.ui-spinner-vertical .ui-spinner-input {
    padding: 2.857rem 0;
}

body .ui-spinner.ui-spinner-vertical .ui-spinner-button {
    height: 2.357rem;
}

body .ui-spinner.ui-spinner-vertical .ui-spinner-button.ui-button.ui-spinner-up {
    border-radius: 4px 4px 0 0;
}

body .ui-spinner.ui-spinner-vertical .ui-spinner-button.ui-button.ui-spinner-down {
    border-radius: 0 0 4px 4px;
}
```

### CODE REVIEW

**Comment by jepsar** [comment]
@melloware WDYT, should I change it to `[_______|-|+]`?

**Comment by melloware** [comment]
2 things...
1) I like the way you have it now with the - on one side an the plus on the other side.
2) The only thing I am not sure about its "horizontal" mode.  When I think Horizontal and Vertical I think the component is drawn horizontally or vertically.  So the only thing I am not sure about is the actual name of the mode.  Anyone else have any thoughts?

**Comment by jepsar** [comment]
Some ideas: buttonsLeftAndRight, leftAndRight, zoom(Mode), scale(Mode), minusPlus(Mode)

**Comment by jepsar** [comment]
Or `buttons="leftAndRight"` (default is `buttons="normal"`)

**Comment by melloware** [comment]
how about maybe `mode="spinner"` which would be the default and `mode="zoom"` for the plus minus buttons?

**Comment by jepsar** [comment]
> @melloware WDYT, should I change it to `[_______|-|+]`?

This could be "zoomRight"

**Comment by stolp** [comment]
> @melloware WDYT, should I change it to `[_______|-|+]`?

My personal unasked for opinion:

I like this: `[_______|-|+]`
and I like this: `[<|_______|>]`

**Comment by jepsar** [comment]
@stolp I can make the icons configurable and add a mode for both the button positions

**Comment by stolp** [comment]
I think this would be great.

**Comment by jepsar** [comment]
By the way, this is what they are doing in PrimeNG. Let's align with that as much as possible.

https://primefaces.org/primeng/showcase/#/inputnumber

Screenshot 2021-05-10 at 12 24 33

Vertical might be something to do later.

**Comment by melloware** [comment]
I agree PrimeTek likes when we align with the other libs and try and even use similar or same attribute names etc.

**Comment by jepsar** [comment]
I've pushed the changes and listed the theme changes. See https://github.com/primefaces/primefaces/pull/7304#issue-634530905

**Comment by jepsar** [comment]
Threw in vertical as well

**Comment by melloware** [comment]
This looks much better!  I really like these changes.

**Review by melloware — APPROVED** [review]

File: src/main/java/org/primefaces/component/spinner/SpinnerRenderer.java [hunk 1]
```diff
@@ -124,12 +123,52 @@ protected void encodeMarkup(FacesContext context, Spinner spinner) throws IOExce
...
+                .add(fixedStyleClass)
+                .add(styleClass)
+                .add(!valid, "ui-state-error")
+                .build();
+    }
+
+    protected String getIconClass(FacesContext context, String custom, String fallback) {
+        return getStyleClassBuilder(context)
+                .add(Spinner.ICON_BASE_CLASS)
+                .add(Optional.ofNullable(custom).orElse(fallback))
```
Comment (tandraschko) [inline src/main/java/org/primefaces/component/spinner/SpinnerRenderer.java]: why not just a simple if? Hope we dont use it in other places

Comment (jepsar) [reply src/main/java/org/primefaces/component/spinner/SpinnerRenderer.java]: @tandraschko Removed the `Optional`. It's not used elsewhere.

**Comment by tandraschko** [comment]
asigned to mert because of CSS
everything else looks good

**Comment by mertsincan** [comment]
Thanks a lot for your contribution! 

Best Regards,

**Comment by jepsar** [comment]
@mertsincan did you read the notes in https://github.com/primefaces/primefaces/pull/7304#issue-634530905 ? This will require a bit of work from your side.

**Pull request merged**